---
output: html_document
editor_options: 
  chunk_output_type: console
---
# CGE Tools Summary

```{r}
#............
# imports
#..........
library(tidyverse)
library(DT)

```

## Goal 
**Overall**: To detect putative drug resistance mutation(s) in clinical isolates of _S. epidermidis_ from a patient with phenotypic resistance.   
**Immediate**: Use the CGE BAP (using plasmid finder seperately due to build issues) to determine change in virulence/plasmid factors

```{r}

smplvls <- c("Sepi01", "Sepi02", "Sepi03", "Sepi04", "Sepi05a", "Sepi05b", "Sepi06", "Sepi07")

paths <- dir(path = "~/Documents/MountPoints/mountedMeshnick/Projects/Sepi_Res_CaseStudy/CGEtools_noplasmidfinder/", pattern = "Sepi", full.names = T)
paths <- paths[!grepl(".out", paths)] # I created this via snakemake

read_cge <- function(path){
  ret <- read_tsv(paste0(path, "/out.tsv")) %>% 
    dplyr::mutate(smpl = basename(path),
                  smpl = stringr::str_split_fixed(smpl, "_", n=2)[,1]) %>% 
    dplyr::select(c("smpl", dplyr::everything()))
    
}



cgetable <- lapply(paths, read_cge) %>% 
  dplyr::bind_rows(.) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = smplvls)) %>% 
  dplyr::arrange(smpl) %>% 
  dplyr::select(c("smpl", "genome_size", "contigs", "n50", 
"species", "mlst", "mlst_genes", "resistance_genes")) # CGE provides data for all, even though I didn't ask for some of these modules 



# the order is inconsistent among the mlst genes, going to reorder
# https://stackoverflow.com/questions/2403122/regular-expression-to-extract-text-between-square-brackets

cgetable.fx <- cgetable %>% 
  dplyr::mutate(
    mlst_genes = stringr::str_extract(mlst_genes, "(?<=\\[).+?(?=\\])") 
    ) 

sortgenes <- function(genes){
  go <- unlist( stringr::str_split(string = genes, pattern = ",") )
  go <- sort(go)
  go <- paste(go, collapse = ", ")
  return(go)
}

cgetable.fx$mlst_genes <- sapply(cgetable.fx$mlst_genes, sortgenes)
cgetable.fx$resistance_genes <- sapply(cgetable.fx$resistance_genes, sortgenes)

```

```{r, results='asis'}
DT::datatable(cgetable.fx, extensions = 'Buttons', 
              options = list(dom = 'Bfrtip',buttons = c('csv'), pageLength=8))
  
```


### PlasmidFinder
The original `CGEtools BAP` on our cluster was hitting a bug when going through plasmidfinder. Cloned locally and ran the docker there. 

```{r}
library(jsonlite)

plsmddirs <- dir(path = "~/Documents/GitHub/Sepi_Res_CaseStudy/plasmidfinder_tools/plasmidfinder/fastq/", 
                 pattern = "Sepi", full.names = T)
plsmddirs <- plsmddirs[!grepl("\\.", plsmddirs)] 
plsmdpaths <- paste0(plsmddirs, "/data.json")

# name these for list later
names(plsmdpaths) <- stringr::str_extract(plsmdpaths, "(?<=fastq//).+?(?=\\_)")

read_plasmidfinder <- function(path){
  ret <- jsonlite::fromJSON(txt = readLines(path)) 
  # handle list structure -- obviously not exportable
  ret <- ret$plasmidfinder$results$`Gram Positive`$gram_positive
  ret <- lapply(ret, as.data.frame) %>% 
    bind_rows(., .id = "name")
  
  return(ret)
}

plsmdfinder <- lapply(plsmdpaths, read_plasmidfinder) %>% 
  dplyr::bind_rows(., .id = "smpl")



plsmdfinder %>% 
  DT::datatable(., extensions = 'Buttons', options = list(dom = 'Bfrtip',buttons = c('csv'), pageLength=20))
  
```


#### Combine Tables for Cleaner Output
Going to drop some of the information from plasmidfinder, and just keep the plasmids that it identified in the final column.
```{r}

plasmids <- plsmdfinder %>% 
  dplyr::group_by(smpl) %>% 
  dplyr::select(c("smpl", "plasmid", "note")) %>%
  dplyr::mutate(plasmidnote = paste0(plasmid, "-", note)) %>% 
  dplyr::select(c("smpl", "plasmidnote")) %>% 
  tidyr::nest() 

plasmids$plasmids <- purrr::map(plasmids$data, function(x){ 
  ret <- unlist(x)
  ret <- sort(ret)
  ret <- paste(ret, collapse = ", ")
  return(ret)
  })

cgetable.fx.plasmids <- plasmids %>% 
  dplyr::select(-c("data")) %>% 
  dplyr::left_join(x = cgetable.fx, y = ., by = "smpl") 




DT::datatable(cgetable.fx.plasmids, 
              extensions = 'Buttons', 
              options = list(dom = 'Bfrtip',buttons = c('csv'), pageLength=20))
  
```

```{r}
save(cgetable.fx, 
     plsmdfinder,
     cgetable.fx.plasmids, 
     file = "~/Documents/GitHub/Sepi_Res_CaseStudy/analyze_sepi_cc/data/derived_data/GEtools_table_outputs.RDA")

```

###### _Session Info_
```{r}
sessionInfo()
```



